src/pool/pool_disjoint.c

```diff
@@ -0,0 +1,1171 @@
+}
+
+bool bucket_can_pool(bucket_t *bucket, bool *to_pool) {
+    size_t new_free_slabs_in_bucket;
+
+    // check if this bucket is used in chunked form or as full slabs
+    bool chunked_bucket = bucket->size <= bucket_chunk_cut_off(bucket);
+    if (chunked_bucket) {
+        new_free_slabs_in_bucket = bucket->chunked_slabs_in_pool + 1;
+    } else {
+        new_free_slabs_in_bucket = bucket->available_slabs_num;
+    }
+
+    if (bucket_capacity(bucket) >= new_free_slabs_in_bucket) {
+        size_t pool_size = 0;
+        utils_atomic_load_acquire(&bucket->shared_limits->total_size,
+                                  &pool_size);
+        while (true) {
+            size_t new_pool_size = pool_size + bucket_slab_alloc_size(bucket);
+
+            if (bucket->shared_limits->max_size < new_pool_size) {
+                break;
+            }
+
+            if (utils_compare_exchange(&bucket->shared_limits->total_size,
+                                       &pool_size, &new_pool_size)) {
+                if (chunked_bucket) {
+                    ++bucket->chunked_slabs_in_pool;
+                }
+
+                bucket_update_stats(bucket, -1, 1);
+                *to_pool = true;
+                return true;
+            }
+        }
+    }
+
+    bucket_update_stats(bucket, -1, 0);
+    *to_pool = false;
+    return false;
+}
+
+utils_rwlock_t *bucket_get_known_slabs_map_lock(bucket_t *bucket) {
+    return &bucket->pool->known_slabs_map_lock;
+}
+
+static size_t size_to_idx(disjoint_pool_t *pool, size_t size) {
+    assert(size <= CutOff && "Unexpected size");
+    assert(size > 0 && "Unexpected size");
+
+    size_t min_bucket_size = (size_t)1 << pool->min_bucket_size_exp;
+    if (size < min_bucket_size) {
+        return 0;
+    }
+
+    // get the position of the leftmost set bit
+    size_t position = getLeftmostSetBitPos(size);
+
+    bool is_power_of_2 = 0 == (size & (size - 1));
+    bool larger_than_halfway_between_powers_of_2 =
+        !is_power_of_2 &&
+        (bool)((size - 1) & ((uint64_t)(1) << (position - 1)));
+    size_t index = (position - pool->min_bucket_size_exp) * 2 +
+                   (int)(!is_power_of_2) +
+                   (int)larger_than_halfway_between_powers_of_2;
+
+    return index;
+}
+
+umf_disjoint_pool_shared_limits_t *
+disjoint_pool_get_limits(disjoint_pool_t *pool) {
+    if (pool->params.shared_limits) {
+        return pool->params.shared_limits;
+    } else {
+        return pool->default_shared_limits;
+    }
+}
+
+bucket_t *disjoint_pool_find_bucket(disjoint_pool_t *pool, size_t size) {
+    size_t calculated_idx = size_to_idx(pool, size);
+
+#ifndef NDEBUG
+    // debug check
+    bucket_t *bucket = pool->buckets[calculated_idx];
+    assert(bucket->size >= size);
+    (void)bucket;
+
+    if (calculated_idx > 0) {
+        bucket_t *bucket_prev = pool->buckets[calculated_idx - 1];
+        assert(bucket_prev->size < size);
+        (void)bucket_prev;
+    }
+#endif // NDEBUG
```

Comment:
If buckets are created on pool creation, do we need to check their sizes and order each time that we are looking for correct one?